crates/re_arrow_store/src/store_event.rs

```diff
@@ -0,0 +1,494 @@
+}
+
+/// Is it an addition or a deletion?
+///
+/// Reminder: ⚠ Do not confuse _a deletion_ and _a clear_ ⚠.
+///
+/// A deletion is the result of a row being completely removed from the store as part of the
+/// garbage collection process.
+///
+/// A clear, on the other hand, is the act of logging an empty [`re_types_core::ComponentBatch`],
+/// either directly using the logging APIs, or indirectly through the use of a
+/// [`re_types_core::archetypes::Clear`] archetype.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum StoreDiffKind {
+    Addition,
+    Deletion,
+}
+
+impl StoreDiffKind {
+    #[inline]
+    pub fn delta(&self) -> i64 {
+        match self {
+            StoreDiffKind::Addition => 1,
+            StoreDiffKind::Deletion => -1,
+        }
+    }
+}
+
+/// Describes an atomic change in the Rerun [`DataStore`]: a row has been added or deleted.
+///
+/// From a query model standpoint, the [`DataStore`] _always_ operates one row at a time:
+/// - The contents of a row (i.e. its columns) are immutable past insertion, by virtue of
+///   [`RowId`]s being unique and non-reusable.
+/// - Similarly, garbage collection always removes _all the data_ associated with a row in one go:
+///   there cannot be orphaned columns. When a row is gone, all data associated with it is gone too.
+///
+/// Refer to field-level documentation for more information.
+#[derive(Debug, Clone, PartialEq)]
+pub struct StoreDiff {
+    /// Addition or deletion?
+    ///
+    /// The store's internals are opaque and don't necessarily reflect the query model (e.g. there
+    /// might be data in the store that cannot by reached by any query).
+    ///
+    /// A [`StoreDiff`] answers a logical question: "does there exist a query path which can return
+    /// data from that row?".
+    pub kind: StoreDiffKind,
+
+    /// What's the row's [`RowId`]?
+    ///
+    /// [`RowId`]s are guaranteed to be unique within a single [`DataStore`].
+    ///
+    /// Put another way, the same [`RowId`] can only appear twice in a [`StoreDiff`] event:
+    /// one addition and (optionally) one deletion (in that order!).
+    pub row_id: RowId,
+
+    /// The [`TimePoint`] associated with that row.
+    ///
+    /// Since insertions and deletions both work on a row-level basis, this is guaranteed to be the
+    /// same value for both the insertion and deletion events (if any).
+    pub timepoint: TimePoint,
+
+    /// The [`EntityPath`] associated with that row.
+    ///
+    /// Since insertions and deletions both work on a row-level basis, this is guaranteed to be the
+    /// same value for both the insertion and deletion events (if any).
+    pub entity_path: EntityPath,
+
+    /// All the [`DataCell`]s associated with that row.
+    ///
+    /// Since insertions and deletions both work on a row-level basis, this is guaranteed to be the
+    /// same set of values for both the insertion and deletion events (if any).
+    pub cells: IntMap<ComponentName, DataCell>,
+}
```

Comment:
love how you clarify all the logical invariants here, will make it much easier for new comers to implement StoreViews!